helix-core/src/main/java/org/apache/helix/spectator/RoutingTableProvider.java

```diff
@@ -376,50 +575,104 @@ public Object removeRoutingTableChangeListener(
+   * Return names of all resources (shown in ExternalView or CustomizedView) in this cluster.
    */
   public Collection<String> getResources() {
-    return _routingTableRef.get().getResources();
+    String key = getRoutingTableKey(DEFAULT_PROPERTY, DEFAULT_TYPE);
+    return _routingTableRefMap.get(key).get().getResources();
+  }
+
+  public Collection<String> getResources(PropertyType propertyType, String type) {
+    String key = getRoutingTableKey(propertyType.name(), type);
+    return _routingTableRefMap.get(key).get().getResources();
+  }
+
+
+  private String getRoutingTableKey(String propertyTypeName, String customizedStateType) {
+    if (propertyTypeName.equals(DEFAULT_PROPERTY) && customizedStateType.equals(DEFAULT_TYPE)) {
+      // Check whether there exist only one snapshot (_routingTableRefMap)
+      if (_routingTableRefMap.keySet().size() == 1) {
+        return _routingTableRefMap.keySet().iterator().next();
+      } else {
+        throw new HelixException("There is none or more than one RoutingTableSnapshot");
+      }
+    }
+
+    if (!propertyTypeName.equals(DEFAULT_PROPERTY) && customizedStateType.equals(DEFAULT_TYPE)) {
+      if (propertyTypeName.equals(PropertyType.CUSTOMIZEDVIEW.name())) {
+        throw new HelixException("Specific type needs to be used for CUSTOMIZEDVIEW PropertyType");
+      }
+    }
+
+    String key = propertyTypeName + "_" + customizedStateType;
+    if (!_routingTableRefMap.containsKey(key)) {
+      throw new HelixException(
+          String.format("Currently there is no snapshot available for PropertyType %s and type %s",
+              propertyTypeName, customizedStateType));
+    }
+    return key;
+
   }
 
   @Override
   @PreFetch(enabled = false)
   public void onExternalViewChange(List<ExternalView> externalViewList,
       NotificationContext changeContext) {
     HelixConstants.ChangeType changeType = changeContext.getChangeType();
-    if (changeType != null && !changeType.getPropertyType().equals(_sourceDataType)) {
+    if (changeType != null && !_sourceDataTypeMap.containsKey(changeType.getPropertyType())) {
       logger.warn(
-          "onExternalViewChange called with mismatched change types. Source data type {}, changed data type: {}",
-          _sourceDataType, changeType);
+          "onExternalViewChange called with mismatched change types. Source data types does not contain changed data type: {}",
+          changeType);
       return;
     }
     // Refresh with full list of external view.
     if (externalViewList != null && externalViewList.size() > 0) {
       // keep this here for back-compatibility, application can call onExternalViewChange directly
       // with externalview list supplied.
-      refresh(externalViewList, changeContext);
+      String keyReference = PropertyType.EXTERNALVIEW.name() + "_" + DEFAULT_TYPE;
+      refresh(externalViewList, Collections.emptyList(), changeContext, keyReference);
```

Comment:
Instead of having one this type of refresh method, we should split it into 2 refresh methods. One for externalview, another for customizeStatem. Maybe the 3rd one for current state.
Otherwise, a typical problem is that as a reviewer, I don't know what is "Collections.emptyList()" for. As a user, I won't be able to know as well.
Moreover, the behavior of passing both parameters is not clean.

Reply:
Sure. Done.